### PR TITLE
Fix character escaping in snippet transformation

### LINKED ISSUE #80394: Snippet transformations: If-type transforms cannot use backslashes
Issue Type: <b>Bug</b>

Steps to Reproduce:

1. Create user snippet:
```json
"Test": {
  "body": [
    "${CURRENT_YEAR/(.+)/${1:+\\\\}/}",
    "${CURRENT_YEAR/(.+)/${1:?\\\\:-}/}",
    "${CURRENT_YEAR/(.+)/${1:-\\\\}/}",
    "${CURRENT_YEAR/(.+)/${1:\\\\}/}",
  ]
}
````
2. Insert snippet into the document.

Expected output:
```
\
\
2019
2019
```
Actual output:
```
${CURRENT_YEAR/(.+)/+\/}
${CURRENT_YEAR/(.+)/+\/}
${CURRENT_YEAR/(.+)/+\/}
${CURRENT_YEAR/(.+)/+\/}
```

This was working correctly in `1.37.x`.
Works correctly if you replace `\\\\` with anything else.

VS Code version: Code 1.38.0 (3db7e09f3b61f915d03bbfa58e258d6eee843f35, 2019-09-03T21:49:13.459Z)
OS version: Windows_NT x64 10.0.17763
Remote OS version: Linux x64 4.4.0-17763-Microsoft
Remote OS version: Linux x64 4.4.0-17763-Microsoft

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i5-7400 CPU @ 3.00GHz (4 x 3000)|
|GPU Status|2d_canvas: enabled<br>flash_3d: enabled<br>flash_stage3d: enabled<br>flash_stage3d_baseline: enabled<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>native_gpu_memory_buffers: disabled_software<br>oop_rasterization: disabled_off<br>protected_video_decode: unavailable_off<br>rasterization: enabled<br>skia_deferred_display_list: disabled_off<br>skia_renderer: disabled_off<br>surface_synchronization: enabled_on<br>video_decode: enabled<br>viz_display_compositor: disabled_off<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|15.91GB (6.26GB free)|
|Process Argv||
|Screen Reader|no|
|VM|0%|

|Item|Value|
|---|---|
|Remote|WSL|
|OS|Linux x64 4.4.0-17763-Microsoft|
|CPUs|Intel(R) Core(TM) i5-7400 CPU @ 3.00GHz (4 x 3000)|
|Memory (System)|15.91GB (6.26GB free)|
|VM|0%|

|Item|Value|
|---|---|
|Remote|WSL|
|OS|Linux x64 4.4.0-17763-Microsoft|
|CPUs|Intel(R) Core(TM) i5-7400 CPU @ 3.00GHz (4 x 3000)|
|Memory (System)|15.91GB (6.26GB free)|
|VM|0%|
</details>

### PR DESCRIPTION
**Pull request by jeanp413**
Fix #80394

### CODE REVIEW

File: src/vs/editor/contrib/snippet/snippetParser.ts [hunk 1]
```diff
@@ -664,27 +664,23 @@ export class SnippetParser {
...
-		let prevToken = <Token>{ type: TokenType.EOF, pos: 0, len: 0 };
-
-		while (this._token.type !== type || prevToken.type === TokenType.Backslash) {
-			if (this._token.type === type) {
-				res += this._scanner.value.substring(pos, prevToken.pos);
-				pos = this._token.pos;
-			}
-			prevToken = this._token;
-			this._token = this._scanner.next();
+		let start = this._token;
```
Comment (ian-sawyer) [inline src/vs/editor/contrib/snippet/snippetParser.ts]: can this be const

File: src/vs/editor/contrib/snippet/snippetParser.ts [hunk 2]
```diff
@@ -664,27 +664,23 @@ export class SnippetParser {
...
+				if (nextToken.type !== TokenType.Dollar
+					&& nextToken.type !== TokenType.CurlyClose
+					&& nextToken.type !== TokenType.Backslash) {
+					return false;
+				}
 			}
+			this._token = this._scanner.next();
 		}
-		res += this._scanner.value.substring(pos, this._token.pos);
+		let value = this._scanner.value.substring(start.pos, this._token.pos).replace(/\\(\$|}|\\)/g, '$1');
```
Comment (ian-sawyer) [inline src/vs/editor/contrib/snippet/snippetParser.ts]: can this be const

**Review by jrieken — APPROVED** [review]
lgtm

**Comment by jrieken** [comment]
Thanks @jeanp413

**Pull request merged**